src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+        KJ_CASE_ONEOF(pending, Pending) {
+          // If we are in the pending state, start the port if we have listeners.
+          // This is technically not spec compliant, but it is what Node.js
+          // supports. Specifically, adding a new message listener using the
+          // addEventListener method is *technically* not supposed to start
+          // the port but we're going to do what Node.js does.
+          if (count > 0 || onmessageValue != kj::none) {
+            start(js);
+          }
+        }
+        KJ_CASE_ONEOF(started, Started) {
+          // If we are in the started state, stop the port if there are no listeners.
+          if (count == 0 && onmessageValue == kj::none) {
+            state = Pending();
+          }
+        }
+        KJ_CASE_ONEOF(_, Closed) {
+          // Nothing to do. We're already closed so we don't care.
+        }
+      }
+    }
+  });
+}
+
+// Deliver the message to the "message" or "messageerror" event on this port.
+void MessagePort::deliver(
+    jsg::Lock& js, kj::Own<rpc::JsValue::Reader> message, jsg::Ref<MessagePort> port) {
+  auto event = js.tryCatch([&] {
+    jsg::Deserializer deserializer(js, message->getV8Serialized());
+    return MessageEvent::New(
+        js, jsg::JsRef(js, deserializer.readValue(js)), kj::str(), port->addRef(), {});
+  }, [&](jsg::Value exception) {
+    return MessageEvent::NewError(
+        js, jsg::JsRef(js, jsg::JsValue(exception.getHandle(js))), kj::str(), port->addRef(), {});
+  });
+
+  // If any of the message/messageerror event handlers throw,
+  // capture the error and pass it to reportError instead of
+  // propagating up.
+  js.tryCatch([&] { port->dispatchEventImpl(js, kj::mv(event)); }, [&](jsg::Value exception) {
+    auto context = js.v8Context();
+    auto& global =
+        jsg::extractInternalPointer<ServiceWorkerGlobalScope, true>(context, context->Global());
+    global.reportError(js, jsg::JsValue(exception.getHandle(js)));
+  });
+}
+
+// Deliver the message to all the jsrpc remotes we have
+kj::Promise<void> MessagePort::sendToRpc(kj::Own<rpc::JsValue::Reader> message) {
+  KJ_IF_SOME(outputLocks, ioContext.waitForOutputLocksIfNecessary()) {
+    co_await outputLocks;
+  }
+  kj::Vector<kj::Promise<void>> promises;
+  for (rpc::JsMessagePort::Client cap: rpcClients) {
+    auto req = cap.callRequest();
+    req.setData(*message);
+    // TODO(message-port): Removing the port when dropped?
+    promises.add(req.send().ignoreResult());
+  }
+  co_await kj::joinPromises(promises.releaseAsArray());
+}
+
+// Deliver the message to this port, buffering if necessary if the port
+// has not been started. Buffered messages will be delivered when the
+// port is started later.
+void MessagePort::deliverMessage(jsg::Lock& js, rpc::JsValue::Reader value) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(pending, Pending) {
+      // We have not yet started the port so buffer to message.
+      pending.add(capnp::clone(value));
```

Comment:
TODO: Tie this into isolate memory accounting so we don't over-buffer.

Reply:
Does this cover transfer property of postMessage() function? if we transfer, we should move rather than clone.

Reply:
I don't know what you mean by "cover transfer property of postMessage()"